lib/src/shared/keepalive.dart

```diff
@@ -0,0 +1,254 @@
+import 'dart:async';
+
+import 'package:clock/clock.dart';
+import 'package:meta/meta.dart';
+
+/// Options to configure a gRPC client for sending keepalive signals.
+class ClientKeepAliveOptions {
+  /// How often a ping should be sent to keep the connection alive.
+  final int? keepaliveTimeMs;
+
+  /// How long the connection should wait before shutting down after no response
+  /// to a ping.
+  final int keepaliveTimeoutMs;
+
+  /// If a connection with no active calls should be kept alive by sending
+  /// pings.
+  final bool keepalivePermitWithoutCalls;
+
+  Duration? get keepaliveTime =>
+      keepaliveTimeMs != null ? Duration(milliseconds: keepaliveTimeMs!) : null;
+
+  Duration get keepaliveTimeout => Duration(milliseconds: keepaliveTimeoutMs);
+
+  const ClientKeepAliveOptions({
+    this.keepaliveTimeMs,
+    this.keepaliveTimeoutMs = 20000,
+    this.keepalivePermitWithoutCalls = false,
+  });
+  bool get sendPings => keepaliveTime != null;
+}
+
+/// Options to configure a gRPC server for receiving keepalive signals.
+class ServerKeepAliveOptions {
+  final int? maxBadPings;
+  final int? minIntervalBetweenPingsWithoutDataMs;
+
+  Duration get minIntervalBetweenPingsWithoutData =>
+      Duration(milliseconds: minIntervalBetweenPingsWithoutDataMs!);
+
+  const ServerKeepAliveOptions({
+    this.minIntervalBetweenPingsWithoutDataMs = 300000,
+    this.maxBadPings = 2,
+  });
+}
+
+enum _ClientKeepAliveState {
+  /// Transport has no active rpcs. We don't need to do any keepalives.
+  idle,
+
+  /// We have scheduled a ping to be sent in the future. We may decide to delay
+  /// it if we receive some data.
+  pingScheduled,
+
+  /// We need to delay the scheduled keepalive ping.
+  pingDelayed,
+
+  /// The ping has been sent out. Waiting for a ping response.
+  pingSent,
+
+  /// Transport goes idle after ping has been sent.
+  idleAndPingSent,
+
+  /// The transport has been disconnected. We won't do keepalives any more.
+  disconnected;
+}
+
+/// A keep alive "manager", deciding when to send pings or shutdown based on the
+/// [ClientKeepAliveOptions].
+class ClientKeepAlive {
```

Comment:
This class can benefit from some form of state transition diagram - otherwise it is very hard to follow how states transition into each other. 

Additionally it is rather confusing which of the `pingTimer` and `pingTimer` can be active at any given state moment. I wonder if there is a better way to express it, e.g. by gluing timers and states together. 

Maybe (just a sketch, maybe some other way to express the same is possible):

```dart
sealed class ClientKeepAliveState {
  void detach();
}

final class StatePingSent extends ClientKeepAliveState {
  final Timer shutdown;
  bool idle;
  
  void detach() {
    shutdown.cancel();
  }
  
  ClientKeepAliveState handle(Event event) => switch (event) {
    Event.frame when !idle => StatePingScheduled(...),
    Event.frame when idle => StateIdle(shutdown: Timer(...)), // inherit the shutdown
  };
} 

final class StatePingScheduled extends ClientKeepAliveState {
  final Timer ping;
} 

class ClientKeepAlive {
  void _handle(Event e) {
    final newState = _state.handle(e);
    if (!identical(newState, _state)) {
      _state.detach();
    }
    _state = newState;
  }

  void onFrameReceived() {
    _handle(Event.frame);  
  }
}
```

Maybe this will end up extremely verbose (probably), but at least it would make a lot of transitions very easy to reason about.

Reply:
I tried something of the sort before, but I found it even more confusing due to the large amount of code needed. But I will try if I can find another solution.